amazon/ion/ioncmodule.c

```diff
@@ -384,48 +409,37 @@ static iERR ionc_write_sequence(hWRITER writer, PyObject* sequence, PyObject* tu
  */
 static iERR ionc_write_struct(hWRITER writer, PyObject* map, PyObject* tuple_as_sexp) {
     iENTER;
-    PyObject * list = PyMapping_Items(map);
-    PyObject * seq = PySequence_Fast(list, "expected a sequence within the map.");
-    PyObject * key = NULL, *val = NULL, *child_obj = NULL;
-    Py_ssize_t len = PySequence_Size(seq);
-    Py_ssize_t i;
-
-    for (i = 0; i < len; i++) {
-        child_obj = PySequence_Fast_GET_ITEM(seq, i);
-        key = PyTuple_GetItem(child_obj, 0);
-        val = PyTuple_GetItem(child_obj, 1);
-        Py_INCREF(child_obj);
-        Py_INCREF(key);
-        Py_INCREF(val);
-
-        if (PyUnicode_Check(key)) {
-            ION_STRING field_name;
-            ion_string_from_py(key, &field_name);
-            IONCHECK(ion_writer_write_field_name(writer, &field_name));
+    PyObject *store = NULL, *key = NULL, *val_list = NULL, *val = NULL;
+    Py_ssize_t pos = 0, i, list_len;
+    if (PyDict_Check(map)) {
+        while (PyDict_Next(map, &pos, &key, &val)) {
+            IONCHECK(write_struct_field(writer, key, val, tuple_as_sexp));
         }
-        else if (key == Py_None) {
-            // if field_name is None, write symbol $0 instead.
-            IONCHECK(_ion_writer_write_field_sid_helper(writer, 0));
+    } else {
+        store = PyObject_CallMethod(map, "_get_store", NULL);
```

Comment:
Just an FYI that you could do: `PyObject_GetAttr(map, "_IonPyDict__store")`

Reply:
> Just an FYI that you could do: `PyObject_GetAttr(map, "_IonPyDict__store")`

This caught my eye, so I found the relevant documentation for anyone else who's curious: [Private name mangling](https://docs.python.org/3/reference/expressions.html#atom-identifiers)

Reply:
This is a better strategy to get `_ _store` attribute. By avoiding calling `PyObject_CallMethod(map, "_get_store", NULL);`, we gained another _**9%**_ performance improvements! After rerun the benchmark before and after changes, there is overall **_40%_** improvement on writing `service_log_legacy` data.
| name                                      |   file_size(B) |   time_min(ns) |   time_mean(ns) |   memory_usage_peak(B) |
|:------------------------------------------|---------------:|---------------:|----------------:|-----------------------:|
| Before |       21270622 |  1737683900.00 |   1848320107.78 |               42584810 |
| After |       21270622 |  1059837258.20 |   1107548862.48 |               42571174 |